Adds filters to get the loading optimization attributes

pereirinha opened the pull request:
This PR adds 2 filters.

The first one, `wp_pre_get_loading_optimization_attributes`, allows short-circuiting the core logic, whereas `wp_get_loading_optimization_attributes` allows filtering the core logic.

Trac ticket: https://core.trac.wordpress.org/ticket/58893

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Commit:
Adds filters to get the loading optimization attributes

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5661,6 +5661,27 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 			}
 	}
 
+	/**
+	 * Filters whether to short-circuit loading optimization attributes.
+	 *
+	 * Returning an array from the filter will effectively short-circuit the loading of optimization attributes,
+	 * returning that value instead.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param false|array $loading_attrs The loading optimization attributes.
+	 * @param array       $attr          Array of the attributes for the tag.
+	 * @param string      $context       Context for the element for which the loading optimization attribute is requested.
+	 */
+	$loading_attrs = apply_filters( 'wp_pre_get_loading_optimization_attributes', false, $attr, $context );
+
+	if ( is_array( $loading_attrs ) ) {
+		return $loading_attrs;
+	}
+
+	// Resets the loading attributes.
+	$loading_attrs = array();
```
normal filters like this are at the state of the function. There is mulitple return statements at the state of this function, meaning this would not fire.

felixarntz replied:
I agree this filter should be moved to the very top of the function, since that would bring several benefits such as:
* A plugin could support more tags than `img` and `iframe`.
* A plugin could add optimization logic for the `template` context (which is for the overall block template).
* A plugin could add optimization logic that is not tied to `width` and `height` attributes (which in fact already applies to `decoding="async"` (see #4991).

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5661,6 +5661,27 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param false|array $loading_attrs The loading optimization attributes.
+	 * @param array       $attr          Array of the attributes for the tag.
+	 * @param string      $context       Context for the element for which the loading optimization attribute is requested.
+	 */
+	$loading_attrs = apply_filters( 'wp_pre_get_loading_optimization_attributes', false, $attr, $context );
```
Two points of feedback here:
* The `$tag_name` value is missing and needs to be passed to the filter as it is critical to make loading optimization changes.
* It's more common in core to use the full function name as filter name and prefix that with "pre_", i.e. `pre_wp_` instead of `wp_pre_`.
```suggestion
	 * @param false|array $loading_attrs The loading optimization attributes.
	 * @param string      $tag_name      The tag name.
	 * @param array       $attr          Array of the attributes for the tag.
	 * @param string      $context       Context for the element for which the loading optimization attribute is requested.
	 */
	$loading_attrs = apply_filters( 'pre_wp_get_loading_optimization_attributes', false, $tag_name, $attr, $context );
```

mukeshpanchal27 replied:
Addition to Felix changes we should add`$loading_attrs` as array instead of `false` otherwise it return bool.

```
* @param array $loading_attrs The loading optimization attributes.
```

felixarntz replied:
@mukeshpanchal27 I don't think that's correct. For a pre-filter, we need to use a clearly different type. If someone wants to override this to return an empty array, they should be able to do so, which with the new change is no longer possible. So I think this should be changed back to how it was, using `false|array` and `false` as the default.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5801,6 +5822,17 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array  $loading_attrs The loading optimization attributes.
+	 * @param array  $attr          Array of the attributes for the tag.
+	 * @param string $context       Context for the element for which the loading optimization attribute is requested.
+	 */
+	$loading_attrs = apply_filters( 'wp_get_loading_optimization_attributes', $loading_attrs, $attr, $context );
```
See above.
```suggestion
	 * @param array  $loading_attrs The loading optimization attributes.
	 * @param string $tag_name      The tag name.
	 * @param array  $attr          Array of the attributes for the tag.
	 * @param string $context       Context for the element for which the loading optimization attribute is requested.
	 */
	$loading_attrs = apply_filters( 'wp_get_loading_optimization_attributes', $loading_attrs, $tag_name, $attr, $context );
```

felixarntz reviewed (changes requested):
@pereirinha Left a bit of feedback below, but otherwise this looks good in terms of production code.

We should also add a few unit tests though. While adding filters is quite basic, let's add least add 1-2 tests covering each filter specifically.

mukeshpanchal27 commented:
@pereirinha, could you kindly address the feedback provided above and rebase the PR? It's worth noting that https://core.trac.wordpress.org/ticket/58894 was merged yesterday, so this may require some adjustments.

mukeshpanchal27 commented:
@pereirinha Could you please update code and unit tests as suggested below?

Add `pre_wp_get_loading_optimization_attributes` filter code after line 5628. and remove your changes for this filter.

```
/**
 * Filters whether to short-circuit loading optimization attributes.
 *
 * Returning an array from the filter will effectively short-circuit the loading of optimization attributes,
 * returning that value instead.
 *
 * @since 6.4.0
 *
 * @param array  $loading_attrs The loading optimization attributes.
 * @param string $tag_name      The tag name.
 * @param array  $attr          Array of the attributes for the tag.
 * @param string $context       Context for the element for which the loading optimization attribute is requested.
 */
$loading_attrs = apply_filters( 'pre_wp_get_loading_optimization_attributes', $loading_attrs, $tag_name, $attr, $context );

if ( ! empty( $loading_attrs ) ) {
	return $loading_attrs;
}
```

**Unit tests:**
```
/**
 * Tests for pre_wp_get_loading_optimization_attributes filter.
 *
 * @ticket 58893
 */
public function test_pre_wp_get_loading_optimization_attributes_filter() {
	add_filter(
		'pre_wp_get_loading_optimization_attributes',
		static function ( $loading_attrs ) {
			$loading_attrs['fetchpriority'] = 'high';
			return $loading_attrs;
		},
		10,
		1
	);

	$attr = $this->get_width_height_for_high_priority();

	$this->assertSame(
		array( 'fetchpriority' => 'high' ),
		wp_get_loading_optimization_attributes( 'img', $attr, 'the_content' ),
		'The filter did not return early fetchpriority attribute'
	);

	// Clean up the filter.
	add_filter( 'pre_wp_get_loading_optimization_attributes', '__return_empty_array' );

	// Modify the loading attributes with any custom attributes.
	add_filter(
		'pre_wp_get_loading_optimization_attributes',
		static function ( $loading_attrs ) {
			$loading_attrs['custom_attr'] = 'custom_value';
			return $loading_attrs;
		},
		10,
		1
	);

	$this->assertSame(
		array( 'custom_attr' => 'custom_value' ),
		wp_get_loading_optimization_attributes( 'img', $attr, 'the_content' ),
		'The filter did not return custom attributes.'
	);
}

/**
 * Tests for wp_get_loading_optimization_attributes filter.
 *
 * @ticket 58893
 */
public function test_wp_get_loading_optimization_attributes_filter() {
	$attr = $this->get_width_height_for_high_priority();

	$this->assertSame(
		array( 'loading' => 'lazy' ),
		wp_get_loading_optimization_attributes( 'img', $attr, 'the_content' ),
		'Before the filter it will not return the loading attribute.'
	);

	add_filter(
		'wp_get_loading_optimization_attributes',
		static function ( $loading_attrs ) {
			unset( $loading_attrs['loading'] );
			$loading_attrs['fetchpriority'] = 'high';
			return $loading_attrs;
		},
		10,
		1
	);

	$this->assertSame(
		array( 'fetchpriority' => 'high' ),
		wp_get_loading_optimization_attributes( 'img', $attr, 'the_content' ),
		'After the filter it will not return the fetchpriority attribute.'
	);
}
```

Commit:
Add `pre_wp_get_loading_optimization_attributes` filter to allow short-circut `wp_get_loading_optimization_attributes()`

Commit:
Include the `$tag_name` on `wp_get_loading_optimization_attributes` filter

Commit:
Add tests for wp_get_loading_optimization_attributes filters

pereirinha reviewed (commented):
@spacedmonkey @mukeshpanchal27 @felixarntz 

Thanks for the input on this one. The PR is updated with your feedback and ready for you to circle back.

Commit:
Fix attributes order

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5627,6 +5627,25 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array  $loading_attrs The loading optimization attributes.
+	 * @param string $tag_name      The tag name.
+	 * @param array  $attr          Array of the attributes for the tag.
+	 * @param string $context       Context for the element for which the loading optimization attribute is requested.
+	 */
+	$loading_attrs = apply_filters( 'pre_wp_get_loading_optimization_attributes', $loading_attrs, $tag_name, $attr, $context );
+
+	if ( ! empty( $loading_attrs ) ) {
```
See https://github.com/WordPress/wordpress-develop/pull/5176#discussion_r1330580812, the filtered type was initially correct and IMO should be reverted to that.
```suggestion
	 * @param array|false  $loading_attrs False by default, or array of loading optimization attributes to short-circuit.
	 * @param string       $tag_name      The tag name.
	 * @param array        $attr          Array of the attributes for the tag.
	 * @param string       $context       Context for the element for which the loading optimization attribute is requested.
	 */
	$loading_attrs = apply_filters( 'pre_wp_get_loading_optimization_attributes', false, $tag_name, $attr, $context );

	if ( is_array( $loading_attrs ) ) {
```

pereirinha replied:
Thanks, @felixarntz for the clarification here. I was pretty sure I'd seen this pattern before.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5627,6 +5627,25 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+	if ( ! empty( $loading_attrs ) ) {
+		return $loading_attrs;
+	}
+
```
Based on the above, we should then initialize `$loading_attrs = array();` here.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5627,6 +5627,25 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 
 	$loading_attrs = array();
```
See other comments, this can be removed (or rather initialized _after_ applying the filter).

spacedmonkey replied:
This variable needs to be moved below the filter.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5789,6 +5808,18 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+	 * @param array  $attr          Array of the attributes for the tag.
+	 * @param string $context       Context for the element for which the loading optimization attribute is requested.
+	 */
+	$loading_attrs = apply_filters( 'wp_get_loading_optimization_attributes', $loading_attrs, $tag_name, $attr, $context );
+
 	return $loading_attrs;
```
This filter looks good, however we will also need to apply it everywhere else further above where there are early returns (except if the "pre-" filter was used). In those instances, you can refer to the filter docs here with the typical `/** This filter is documented in wp-includes/media.php */` annotation.

Small nit-pick, we can avoid the variable:
```suggestion
	return apply_filters( 'wp_get_loading_optimization_attributes', $loading_attrs, $tag_name, $attr, $context );
```

pereirinha replied:
Good point in the early bails.
Having the filter being applied to a variable, comes from a personal preference, as it's easy to add a breakpoint and see if the filter is changing the value.

felixarntz commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,86 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+			wp_get_loading_optimization_attributes( 'img', $attr, 'the_content' ),
+			'The filter did not return custom attributes.'
+		);
+	}
```
Per the above, this test will need minor modifications given the pre-filter should initially use `false`, not already an array.

felixarntz reviewed (commented):
@pereirinha This looks almost good to go, however the pre filter IMO was correct before in how the typing was used. Right now it's limited to returning a non-empty array, but it should also be possible to short-circuit with an empty array since not every tag can be expected to get loading optimization attributes.

The other thing is that the post filter will also need to be applied in all early returns in the function, except the one where the pre filter is applied.

pereirinha committed:
Update src/wp-includes/media.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

pereirinha committed:
Update src/wp-includes/media.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

Commit:
Initialize `$loading_attrs` after the pre filter

Commit:
Filter the early bails for the `wp_get_loading_optimization_attributes()`

Commit:
Update tests to account for the new default for the `pre_wp_get_loading_optimization_attributes` filter

pereirinha commented:
Thanks again, @mukeshpanchal27, @felixarntz, and @spacedmonkey, for the feedback.

I addressed all the comments so far.
This is ready for you to circle back.

mukeshpanchal27 commented on src/wp-includes/media.php:
```diff
@@ -5625,24 +5625,46 @@ function wp_get_webp_info( $filename ) {
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array|false  $loading_attrs False by default, or array of loading optimization attributes to short-circuit.
+	 * @param string       $tag_name      The tag name.
+	 * @param array        $attr          Array of the attributes for the tag.
+	 * @param string       $context       Context for the element for which the loading optimization attribute is requested.
```
```suggestion
	 * @param array|false $loading_attrs False by default, or array of loading optimization attributes to short-circuit.
	 * @param string      $tag_name      The tag name.
	 * @param array       $attr          Array of the attributes for the tag.
	 * @param string      $context       Context for the element for which the loading optimization attribute is requested.
```

Remove extra spaces

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,101 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+		add_filter(
+			'pre_wp_get_loading_optimization_attributes',
+			static function ( $loading_attrs ) {
+				$loading_attrs['fetchpriority'] = 'high';
```
```suggestion
				if ( false === $loading_attrs ) {
					// Initialize as an empty array.
					$loading_attrs = array();
				}
				$loading_attrs['fetchpriority'] = 'high';
```

As per the comment by @felixarntz on https://github.com/WordPress/wordpress-develop/pull/5176#discussion_r1330596737, it is necessary to incorporate this adjustment to prevent the occurrence of the `Automatic conversion of false to array is deprecated` error in PHP 8.1 and subsequent versions.

mukeshpanchal27 replied:
Check https://github.com/WordPress/wordpress-develop/actions/runs/6242000649/job/16945177959?pr=5176 for failaur.

pereirinha replied:
Thanks @mukeshpanchal27. Either I missed that comment, or it was not clear to me. I think it was the latter, as I believed that Felix was mentioning the tests for the case of providing the false and the empty array as values to that filter.

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,101 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+		add_filter(
+			'pre_wp_get_loading_optimization_attributes',
+			static function ( $loading_attrs ) {
+				$loading_attrs['custom_attr'] = 'custom_value';
```
```suggestion
				if ( false === $loading_attrs ) {
					// Initialize as an empty array.
					$loading_attrs = array();
				}
				$loading_attrs['custom_attr'] = 'custom_value';
```

See above.

mukeshpanchal27 reviewed (changes requested):
Thank you, @pereirinha, for the updates. The unit tests are failing on PHP 8.1 and later versions. 

I have provided feedback that addresses the PHP error. I am planning to approve it, and once those changes are incorporated, the pull request will be ready for the final review from @felixarntz.

pereirinha committed:
Update src/wp-includes/media.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pereirinha committed:
Update tests/phpunit/tests/media.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pereirinha committed:
Update tests/phpunit/tests/media.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

pereirinha commented:
Thanks, @mukeshpanchal27, for the clarification on the outstanding issue.
I merged your code, and I'm running tests locally to confirm, but this should now be ready.

mukeshpanchal27 reviewed (approved):
Thanks @pereirinha for the update. Changes look good to me.

spacedmonkey commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,109 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+		// Clean up the filter.
+		add_filter( 'pre_wp_get_loading_optimization_attributes', '__return_false' );
+
+		$this->assertSame(
```
```suggestion
		$this->assertSameSets(
```

spacedmonkey commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,109 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+		// Return no loading attributes.
+		add_filter( 'pre_wp_get_loading_optimization_attributes', '__return_empty_array' );
+
+		$this->assertSame(
```
```suggestion
		$this->assertSameSets(
```

spacedmonkey commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,109 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+			1
+		);
+
+		$this->assertSame(
```
```suggestion
		$this->assertSameSets(
```

spacedmonkey commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,109 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+			1
+		);
+
+		$this->assertSame(
```
```suggestion
		$this->assertSameSets(
```

spacedmonkey commented on tests/phpunit/tests/media.php:
```diff
@@ -5440,6 +5440,109 @@ public function test_wp_get_loading_optimization_attributes_image_before_loop_in
+	public function test_wp_get_loading_optimization_attributes_filter() {
+		$attr = $this->get_width_height_for_high_priority();
+
+		$this->assertSame(
```
```suggestion
		$this->assertSameSets(
```

spacedmonkey reviewed (changes requested):
One small piece of feedback. But this is really close.

pereirinha commented:
> One small piece of feedback. But this is close.

Thanks @spacedmonkey. I haven't encountered the `assertSameSets` before, but it is excellent for the other ticket I'm working on.

For this case, given that we are expecting only one key in the array, I wonder if it will make any real difference.

spacedmonkey commented:
> For this case, given that we are expecting only one key in the array, I wonder if it will make any real difference.

If that is the case we should something like this. 

```
$attrs = wp_get_loading_optimization_attributes( 'img', $attr, 'the_content' );
$this->assertIsArray( $attrs );
$this->assertArrayHasKey( 'custom_attr', $attrs );
$this->assertSame( 'custom_attr', $filtered_widgets['custom_attr'] );
```

felixarntz committed:
Use assertSameSets() instead of assertSame() on associative arrays.

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

felixarntz commented:
@spacedmonkey @pereirinha Apologies, I didn't see your discussion here and just applied the suggestions. While I agree it's not a real issue here, using `assertSameSets` is generally a bit cleaner when doing assertions on associative arrays, since the order doesn't matter. So for one item it doesn't make a difference, but it's still "more correct" to use `assertSameSets` just because it is an associative array - e.g. in case the tests were to be adjusted later.

Commit:
Merge branch 'trunk' into filter/optimization-attributes

felixarntz reviewed (approved):
Thank you for the updates @pereirinha, LGTM!

spacedmonkey reviewed (approved):
I would like to make those unit test changes. But we can do that after this is commited

pereirinha commented:
I got some notifications that Felix had merged https://github.com/WordPress/wordpress-develop/commit/896b25e63fa057e7902613ba17a3789338f1d4e5. I thought this was revolving everything outstanding.

Is there still an action item pending here?

felixarntz commented:
@pereirinha All good, I'm about to commit this.

Once I have committed it, can you refresh your branch from #4991 with the latest `trunk` to integrate the changes, in case there are any merge conflicts?

Commit:
Merge branch 'trunk' into filter/optimization-attributes

felixarntz commented:
Committed in https://core.trac.wordpress.org/changeset/56651
